### PR TITLE
soc: arm: efr32bg22: Add support for Gecko SPI

### PR DESCRIPTION
**Pull request by fkokosinski**
This PR adds support for the Gecko SPI and adds an SPI transaction sample in `samples/drivers/spi_efr32bg22/src/spi.c`.

### CODE REVIEW

File: samples/drivers/spi_efr32bg22/src/spi.c [hunk 1]
```diff
@@ -0,0 +1,135 @@
...
+		LOG_ERR("Failed to disable SPI bus, return code: %d", ret);
+	}
+
+	LOG_INF("Done");
+}
+
+void main(void)
+{
+	test_spi();
+}
```
Comment (gmarull) [inline samples/drivers/spi_efr32bg22/src/spi.c:135]: -1 on platform specific samples for interfaces. Also, the SPI enable bit should be part of the SPI driver and be specified in the SPI bindings.

**Review by gmarull — CHANGES_REQUESTED** [review]

File: samples/drivers/spi_efr32bg22/src/spi.c (hunk 1 above)
Comment (lpawelcz) [reply samples/drivers/spi_efr32bg22/src/spi.c]: Removed the sample. Please note that the SPI enable bit was not used as CS line. It was used for driving on-board switch. The switch connected SPI peripheral used in initial SPI driver tests to the common SPI bus on the platform. This should be handled by users in their own applications so I also removed this part from the PR.
FYI: I'm pasting here a figure from Thunderboard EFR32BG22 User Guide that highlights the usage of SPI enable bit (EFR32_IMU_ENABLE)
image

Comment (gmarull) [reply samples/drivers/spi_efr32bg22/src/spi.c]: note that on-board switch should either use the `regulator-fixed` compatible, or, if specific to a device you can also use `-gpios` prop to directly toggle the GPIO in the driver (probably combined with PM).

File: dts/arm/silabs/efr32bg22-pinctrl.dtsi [hunk 2]
```diff
@@ -10,10 +10,9 @@
 	/* configuration for uart0 device, default state */
 	usart0_default: usart0_default {
 		group1 {
-			/* configure PE.11 as UART_RX and PE.10 as UART_TX */
-			psels = <GECKO_PSEL(UART_TX, E, 10)>,
-				<GECKO_PSEL(UART_RX, E, 11)>,
-				<GECKO_LOC(UART, 0)>;
+			psels = <GECKO_PSEL(SPI_SCK, C, 2)>,
+				<GECKO_PSEL(SPI_MOSI, C, 0)>,
+				<GECKO_PSEL(SPI_MISO, C, 1)>;
```
Comment (gmarull) [inline dts/arm/silabs/efr32bg22-pinctrl.dtsi:15]: non-blocking: maybe a comment saying UART is operating as SPI here.

File: dts/arm/silabs/efr32bg22.dtsi [hunk 3]
```diff
@@ -54,11 +54,31 @@
...
+				reg = <0>;
+				spi-max-frequency = <80000000>;
+				size = <0x800000>;
+				jedec-id = [c2 28 14];
+				sfdp-bfp = [
+					 e5 20 f1 ff  ff ff 7f 00  44 eb 08 6b  08 3b 04 bb
+					 ee ff ff ff  ff ff 00 ff  ff ff 00 ff  0c 20 0f 52
+					 10 d8 00 ff  23 72 f5 00  82 ed 04 b7  44 83 38 44
+					 30 b0 30 b0  f7 c4 d5 5c  00 be 29 ff  f0 d0 ff ff
+				];
```
Comment (gmarull) [inline dts/arm/silabs/efr32bg22.dtsi:80]: is this flash embedded in the SoC?

Comment (lpawelcz) [reply dts/arm/silabs/efr32bg22.dtsi:80]: It is a separate on-board chip, I guess I should move this to board-specific dts then

Comment (lpawelcz) [reply dts/arm/silabs/efr32bg22.dtsi]: moved `mx25r80` to `boards/arm/efr32bg_sltb010a/efr32bg_sltb010a.dts`

File: dts/arm/silabs/efr32bg22-pinctrl.dtsi (hunk 2 above)
Comment (lpawelcz) [reply dts/arm/silabs/efr32bg22-pinctrl.dtsi:15]: Added comments

File: samples/drivers/spi_efr32bg22/src/spi.c (hunk 1 above)
Comment (lpawelcz) [reply samples/drivers/spi_efr32bg22/src/spi.c]: I'm not sure if I understand that correctly. Would you like me to reintroduce the SPI enable bit but using existing compatible or the `-gpios` prop? It is specific to one board and one sensor on that board. That is also why I don't think it should be handled by SPI driver. In my opinion this should be handled on application level when user decides to actually power up and use the sensor. Please let me know if I can leave the SPI enable pin removed or if I should rework it.

Comment (gmarull) [reply samples/drivers/spi_efr32bg22/src/spi.c]: If that's the case, you can define a `fixed-regulator` for that GPIO. Then clients can use the `vin-supply` property to capture the regulator handle and use `regulator_enable/disable` as needed. However, usage of regulators is rarely supported in subsystems, I hope this improves in the future.

Comment (lpawelcz) [reply samples/drivers/spi_efr32bg22/src/spi.c]: Added in https://github.com/zephyrproject-rtos/zephyr/pull/53255/commits/7402a6bb6f9fc6a2aa6998f73590f71a957a4c57.

File: dts/arm/silabs/efr32bg22.dtsi [hunk 4]
```diff
@@ -54,11 +54,18 @@
...
+			compatible = "silabs,gecko-spi-usart";
 			reg = <0x5005C000 0x400>;
 			interrupts = <13 0>, <14 0>;
 			interrupt-names = "rx", "tx";
 			status = "disabled";
+
+			#address-cells = <1>;
+			#size-cells = <0>;
+
+			cs-gpios = <&gpioc 3 GPIO_ACTIVE_LOW>;
```
Comment (gmarull) [inline dts/arm/silabs/efr32bg22.dtsi:66]: is this GPIO connected internally at SoC level?

**Review by gmarull — DISMISSED** [review]
one comment

File: dts/arm/silabs/efr32bg22.dtsi (hunk 4 above)
Comment (kgugala) [reply dts/arm/silabs/efr32bg22.dtsi:66]: this refers to SoC level GPIO controller

Comment (gmarull) [reply dts/arm/silabs/efr32bg22.dtsi:66]: this refers to a specific GPIO pin, not only the controller. Is it hard-wired in the SoC?

Comment (lpawelcz) [reply dts/arm/silabs/efr32bg22.dtsi]: CS is connected at the board level - moved `cs-gpios` to board-specific DTS.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
You should add spi to the board yaml and doc, but otherwise LGTM

**Comment by lpawelcz** [comment]
@MaureenHelm I've listed SPI in the docs and in the yaml file with Twister metadata

**Review by MaureenHelm — DISMISSED** [review]

**Comment by lpawelcz** [comment]
Hi @MaureenHelm @gmarull, sorry for pinging you again but I had to rebase this PR to solve conflicts created by merging [I2C support](https://github.com/zephyrproject-rtos/zephyr/pull/53246) PR. Please take another look at this PR.

**Review by MaureenHelm — APPROVED** [review]

**Review by teburd — APPROVED** [review]

**Pull request merged**